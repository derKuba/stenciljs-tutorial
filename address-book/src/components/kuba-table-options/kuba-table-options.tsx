import { Component, Host, h, Element, Prop } from '@stencil/core';

@Component({
  tag: 'kuba-table-options',
  styleUrl: 'kuba-table-options.css',
  shadow: true,
})
export class KubaTableOptions {

  @Element() $el: HTMLKubaTableOptionsElement;
  @Prop() delete;

  private headItems = [];
  private bodyItems = [];

  connectedCallback() {
    const headOptions = Array.from(this.$el.querySelectorAll("kuba-table-options-head option"));
    this.headItems = headOptions.map(item => item.getAttribute("value"));

    const bodyOptions = Array.from(this.$el.querySelectorAll("kuba-table-options-body option"));
    const bodyOptionsValues = bodyOptions.map(item => item.getAttribute("value"));
    this.bodyItems = this.splitArrayIntoChunks(bodyOptionsValues, this.headItems.length);
  }

  // https://ourcodeworld.com/articles/read/278/how-to-split-an-array-into-chunks-of-the-same-size-easily-in-javascript
  private splitArrayIntoChunks = (items, size) => {
    let chunks = [];
    while (items.length) {
      chunks.push(items.splice(0, size));
    }
    return chunks;
  }

  render() {

    return (
      <Host>
        <table>
          <thead>
            <tr>
              {
                this.headItems.length > 0 && this.headItems.map(row => <th>{row}</th>)
              }
            </tr>
          </thead>
          <tbody>
            {
              this.bodyItems.length > 0 && this.bodyItems.map(row => <tr>
                {row.map((cell, index) => {
                  if (index === row.length-1) {
                    return (<td>
                       <stencil-route-link url={`/contact/${row[row.length-1]}`}>
                          editieren
                        </stencil-route-link>
                      <button onClick={()=> this.delete(row[row.length-1])}>l√∂schen</button>
                    </td>)
                  }
                  return <td>{cell}</td>
                })
                }
              </tr>)
            }
          </tbody>
        </table>
      </Host>
    );
  }
}
