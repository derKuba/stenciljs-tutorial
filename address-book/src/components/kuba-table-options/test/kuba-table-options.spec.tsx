import { newSpecPage } from "@stencil/core/testing";
import { KubaTableOptions } from "../kuba-table-options";

describe("kuba-table-options", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [KubaTableOptions],
      html: `<kuba-table-options>
    <kuba-table-options-head>
        <option value="id" />
        <option value="name" />
        <option value="vorname" />
    </kuba-table-options-head>

    <kuba-table-options-body>
        <option value="1" />
        <option value="Max" />
        <option value="Muster" />

        <option value="2" />
        <option value="Maxine" />
        <option value="Muster" />

        <option value="3" />
        <option value="Max" />
        <option value="Muster" />
    </kuba-table-options-body>
</kuba-table-options>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
