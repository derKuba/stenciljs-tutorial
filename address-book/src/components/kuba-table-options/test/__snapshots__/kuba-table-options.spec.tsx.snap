// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kuba-table-options renders 1`] = `
<kuba-table-options>
  <mock:shadow-root>
    <table>
      <thead>
        <tr>
          <th>
            id
          </th>
          <th>
            name
          </th>
          <th>
            vorname
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            Max
          </td>
          <td>
            <stencil-route-link url="/contact/Muster">
              editieren
            </stencil-route-link>
            <button>
              löschen
            </button>
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            Maxine
          </td>
          <td>
            <stencil-route-link url="/contact/Muster">
              editieren
            </stencil-route-link>
            <button>
              löschen
            </button>
          </td>
        </tr>
        <tr>
          <td>
            3
          </td>
          <td>
            Max
          </td>
          <td>
            <stencil-route-link url="/contact/Muster">
              editieren
            </stencil-route-link>
            <button>
              löschen
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </mock:shadow-root>
  <kuba-table-options-head>
    <option value="id"></option>
    <option value="name"></option>
    <option value="vorname"></option>
  </kuba-table-options-head>
  <kuba-table-options-body>
    <option value="1"></option>
    <option value="Max"></option>
    <option value="Muster"></option>
    <option value="2"></option>
    <option value="Maxine"></option>
    <option value="Muster"></option>
    <option value="3"></option>
    <option value="Max"></option>
    <option value="Muster"></option>
  </kuba-table-options-body>
</kuba-table-options>
`;
